### PR TITLE
Trigger callback in script

### LINKED ISSUE #205: Expose triggers to scripts
Allow scripts to hook into triggers to provide functionality.

### PR DESCRIPTION
**Pull request by andens**
Resolves #205

### CODE REVIEW

File: src/Engine/Manager/PhysicsManager.cpp [hunk 1]
```diff
@@ -87,20 +94,34 @@ void PhysicsManager::Update(float deltaTime) {
     }
 }
```
Comment (NinjaDanz3r) [inline src/Engine/Manager/PhysicsManager.cpp]: MINOR The function 'MakeTrigger' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

File: src/Engine/Manager/PhysicsManager.cpp [hunk 2]
```diff
@@ -78,6 +80,11 @@ void PhysicsManager::Update(float deltaTime) {
...
         // Update rotation.
         entity->rotation += physicsComp->angularVelocity * 360.f * deltaTime;
+
+        physicsComp->GetRigidBody().Position(entity->position);
+        dynamicsWorld->removeRigidBody(physicsComp->GetRigidBody().GetRigidBody());
+        dynamicsWorld->addRigidBody(physicsComp->GetRigidBody().GetRigidBody());
+        physicsComp->GetRigidBody().GetRigidBody()->setGravity(btVector3(0, 0, 0));
     }
 
     dynamicsWorld->stepSimulation(deltaTime, 10);
```
Comment (NinjaDanz3r) [inline src/Engine/Manager/PhysicsManager.cpp]: MINOR The function 'MakeRigidBody' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

File: src/Engine/Physics/Trigger.cpp [hunk 3]
```diff
@@ -1,26 +1,30 @@
...
 namespace Physics {
 
-    Trigger::Trigger(Shape* shape) : btCollisionWorld::ContactResultCallback() {
+    Trigger::Trigger(Component::Physics* comp) : btCollisionWorld::ContactResultCallback() {
         trigger = new btCollisionObject();
-        trigger->setCollisionShape(shape->GetShape());
+        
+        trigger->setCollisionShape(comp->GetShape().GetShape());
+        trigger->setWorldTransform(comp->GetRigidBody().GetRigidBody()->getWorldTransform());
     }
```
Comment (NinjaDanz3r) [inline src/Engine/Physics/Trigger.cpp]: MINOR The function 'addSingleResult' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

File: src/Engine/Physics/Trigger.cpp [hunk 4]
```diff
@@ -30,12 +34,12 @@ namespace Physics {
 
     void Trigger::Process(btCollisionWorld& world) {
         for (auto& observerPair : observers) {
-            world.contactPairTest(trigger, observerPair.first->GetRigidBody(), *this);
+            world.contactPairTest(trigger, observerPair.first, *this);
```
Comment (NinjaDanz3r) [inline src/Engine/Physics/Trigger.cpp]: MINOR The function 'OnEnter' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 17 issues
* ![MINOR][MINOR] 17 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L171](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d3f2b9b3e2ca46f3e364c02cb4914346091c4c21/src/Engine/Hymn.cpp#L171): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [PhysicsManager.cpp#L129](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d3f2b9b3e2ca46f3e364c02cb4914346091c4c21/src/Engine/Manager/PhysicsManager.cpp#L129): The function 'GetPhysicsComponents' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L215](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d3f2b9b3e2ca46f3e364c02cb4914346091c4c21/src/Engine/Manager/RenderManager.cpp#L215): Variable 'name' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [RenderManager.cpp#L225](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d3f2b9b3e2ca46f3e364c02cb4914346091c4c21/src/Engine/Manager/RenderManager.cpp#L225): The function 'GetAnimations' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L243](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d3f2b9b3e2ca46f3e364c02cb4914346091c4c21/src/Engine/Manager/RenderManager.cpp#L243): The function 'GetDirectionalLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L282](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d3f2b9b3e2ca46f3e364c02cb4914346091c4c21/src/Engine/Manager/RenderManager.cpp#L282): The function 'GetMaterials' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L300](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d3f2b9b3e2ca46f3e364c02cb4914346091c4c21/src/Engine/Manager/RenderManager.cpp#L300): The function 'GetMeshes' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L320](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d3f2b9b3e2ca46f3e364c02cb4914346091c4c21/src/Engine/Manager/RenderManager.cpp#L320): The function 'GetPointLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L341](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d3f2b9b3e2ca46f3e364c02cb4914346091c4c21/src/Engine/Manager/RenderManager.cpp#L341): The function 'GetSpotLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Engine/Manager/PhysicsManager.cpp [hunk 5]
```diff
@@ -87,20 +94,34 @@ void PhysicsManager::Update(float deltaTime) {
...
+        entity->rotation = glm::eulerAngles(Physics::btToGlm(trans.getRotation()));
+    }
+}
+
+void PhysicsManager::OnTriggerEnter(Component::Physics* triggerBody, Component::Physics* object, std::function<void()> callback) {
+    auto trigger = MakeTrigger(triggerBody);
+    trigger->OnEnter(object, callback);
+}
+
 Physics::RigidBody* PhysicsManager::MakeRigidBody(Physics::Shape* shape, float mass) {
```
Comment (NinjaDanz3r) [inline src/Engine/Manager/PhysicsManager.cpp]: MINOR The function 'MakeRigidBody' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

File: src/Engine/Physics/Trigger.cpp [hunk 6]
```diff
@@ -1,26 +1,30 @@
...
         trigger = new btCollisionObject();
-        trigger->setCollisionShape(shape->GetShape());
+        
+        trigger->setCollisionShape(comp->GetShape().GetShape());
+        trigger->setWorldTransform(comp->GetRigidBody().GetRigidBody()->getWorldTransform());
     }
 
     // Called with each contact for our own processing. This is where we can
     // test whether the given listeners are intersecting the trigger volume.
     btScalar Trigger::addSingleResult(btManifoldPoint& cp,
```
Comment (NinjaDanz3r) [inline src/Engine/Physics/Trigger.cpp]: MINOR The function 'addSingleResult' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 15 issues
* ![MINOR][MINOR] 15 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L171](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/55ae7037f96d61aec7a093737027b245815f2625/src/Engine/Hymn.cpp#L171): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [PhysicsManager.cpp#L166](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/55ae7037f96d61aec7a093737027b245815f2625/src/Engine/Manager/PhysicsManager.cpp#L166): The function 'GetPhysicsComponents' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L216](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/55ae7037f96d61aec7a093737027b245815f2625/src/Engine/Manager/RenderManager.cpp#L216): Variable 'name' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [RenderManager.cpp#L226](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/55ae7037f96d61aec7a093737027b245815f2625/src/Engine/Manager/RenderManager.cpp#L226): The function 'GetAnimations' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L244](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/55ae7037f96d61aec7a093737027b245815f2625/src/Engine/Manager/RenderManager.cpp#L244): The function 'GetDirectionalLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L283](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/55ae7037f96d61aec7a093737027b245815f2625/src/Engine/Manager/RenderManager.cpp#L283): The function 'GetMaterials' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L301](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/55ae7037f96d61aec7a093737027b245815f2625/src/Engine/Manager/RenderManager.cpp#L301): The function 'GetMeshes' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L321](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/55ae7037f96d61aec7a093737027b245815f2625/src/Engine/Manager/RenderManager.cpp#L321): The function 'GetPointLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L342](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/55ae7037f96d61aec7a093737027b245815f2625/src/Engine/Manager/RenderManager.cpp#L342): The function 'GetSpotLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 15 issues
* ![MINOR][MINOR] 15 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L171](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/40d7f835489736ea48b3a4bc82f80e45e6a69b07/src/Engine/Hymn.cpp#L171): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [PhysicsManager.cpp#L166](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/40d7f835489736ea48b3a4bc82f80e45e6a69b07/src/Engine/Manager/PhysicsManager.cpp#L166): The function 'GetPhysicsComponents' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L216](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/40d7f835489736ea48b3a4bc82f80e45e6a69b07/src/Engine/Manager/RenderManager.cpp#L216): Variable 'name' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [RenderManager.cpp#L226](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/40d7f835489736ea48b3a4bc82f80e45e6a69b07/src/Engine/Manager/RenderManager.cpp#L226): The function 'GetAnimations' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L244](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/40d7f835489736ea48b3a4bc82f80e45e6a69b07/src/Engine/Manager/RenderManager.cpp#L244): The function 'GetDirectionalLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L283](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/40d7f835489736ea48b3a4bc82f80e45e6a69b07/src/Engine/Manager/RenderManager.cpp#L283): The function 'GetMaterials' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L301](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/40d7f835489736ea48b3a4bc82f80e45e6a69b07/src/Engine/Manager/RenderManager.cpp#L301): The function 'GetMeshes' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L321](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/40d7f835489736ea48b3a4bc82f80e45e6a69b07/src/Engine/Manager/RenderManager.cpp#L321): The function 'GetPointLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L342](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/40d7f835489736ea48b3a4bc82f80e45e6a69b07/src/Engine/Manager/RenderManager.cpp#L342): The function 'GetSpotLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Engine/Physics/Trigger.hpp [hunk 7]
```diff
@@ -20,7 +23,7 @@ namespace Physics {
             /**
              * @param shape The shape of the trigger volume.
```
Comment (SwedishSock) [inline src/Engine/Physics/Trigger.hpp]: Trigger takes a Component::Physics* comp, change comment to reflect this.

File: src/Engine/Physics/Trigger.hpp [hunk 8]
```diff
@@ -40,11 +43,11 @@ namespace Physics {
             void Process(btCollisionWorld& world);
 
             /// Implementation of ITrigger::OnEnter.
-            virtual void OnEnter(RigidBody& body, std::function<void(RigidBody& body)> observer) override;
+            virtual void OnEnter(Component::Physics* body, std::function<void()> observer) override;
```
Comment (SwedishSock) [inline src/Engine/Physics/Trigger.hpp]: Add Doxygen commentation (parameters)

File: src/Engine/Manager/PhysicsManager.hpp [hunk 9]
```diff
@@ -35,6 +36,12 @@ class PhysicsManager {
          */
         void Update(float deltaTime);
         
+        /// Update transforms of entities according to positions of physics
+        /// components.
+        void UpdateEntityTransforms();
+        
+        void OnTriggerEnter(Component::Physics* triggerBody, Component::Physics* object, std::function<void()> callback);
```
Comment (SwedishSock) [inline src/Engine/Manager/PhysicsManager.hpp]: Add Doxygen comment

File: src/Engine/Manager/PhysicsManager.hpp [hunk 10]
```diff
@@ -35,6 +36,12 @@ class PhysicsManager {
          */
         void Update(float deltaTime);
         
+        /// Update transforms of entities according to positions of physics
+        /// components.
```
Comment (SwedishSock) [inline src/Engine/Manager/PhysicsManager.hpp]: Do you need 2 rows for this?

File: src/Engine/Component/Physics.hpp [hunk 11]
```diff
@@ -35,6 +36,8 @@ namespace Component {
              */
             void SetShape(::Physics::Shape* shape);
             
+            ::Physics::RigidBody& GetRigidBody();
```
Comment (SwedishSock) [inline src/Engine/Component/Physics.hpp]: Doxygen commentation

File: src/Engine/Manager/PhysicsManager.hpp (hunk 10 above)
Comment (andens) [reply src/Engine/Manager/PhysicsManager.hpp]: Mostly for ourselves to not make lines too long. Doxygen output has no line break so it's cool.

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 15 issues
* ![MINOR][MINOR] 15 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L171](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/da7a6417031efcce4386f82a76ec3a23e4b5a83e/src/Engine/Hymn.cpp#L171): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [PhysicsManager.cpp#L166](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/da7a6417031efcce4386f82a76ec3a23e4b5a83e/src/Engine/Manager/PhysicsManager.cpp#L166): The function 'GetPhysicsComponents' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L216](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/da7a6417031efcce4386f82a76ec3a23e4b5a83e/src/Engine/Manager/RenderManager.cpp#L216): Variable 'name' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [RenderManager.cpp#L226](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/da7a6417031efcce4386f82a76ec3a23e4b5a83e/src/Engine/Manager/RenderManager.cpp#L226): The function 'GetAnimations' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L244](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/da7a6417031efcce4386f82a76ec3a23e4b5a83e/src/Engine/Manager/RenderManager.cpp#L244): The function 'GetDirectionalLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L283](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/da7a6417031efcce4386f82a76ec3a23e4b5a83e/src/Engine/Manager/RenderManager.cpp#L283): The function 'GetMaterials' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L301](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/da7a6417031efcce4386f82a76ec3a23e4b5a83e/src/Engine/Manager/RenderManager.cpp#L301): The function 'GetMeshes' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L321](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/da7a6417031efcce4386f82a76ec3a23e4b5a83e/src/Engine/Manager/RenderManager.cpp#L321): The function 'GetPointLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [RenderManager.cpp#L342](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/da7a6417031efcce4386f82a76ec3a23e4b5a83e/src/Engine/Manager/RenderManager.cpp#L342): The function 'GetSpotLights' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**